frappe/deprecation_dumpster.py

```diff
@@ -33,10 +35,72 @@ class Color:
 	CYAN = 96
 
 
+# we use Warning because DeprecationWarning has python default filters which would exclude them from showing
+# see also frappe.__init__ enabling them when a dev_server
 class FrappeDeprecationWarning(Warning):
-	...
+	"""Used as base class when current_version.startswith(graduation)"""
 
 
+class PendingFrappeDeprecationWarning(Warning):
+	"""Used as base class when not current_version.startswith(graduation)"""
+
+
+def _get_dynamic_warning_class_name(graduation: str) -> str:
+	# Scrub the graduation string to ensure it's a valid class name
+	cleaned_graduation = re.sub(r"\W|^(?=\d)", "_", graduation.upper())
+	return f"{cleaned_graduation}FrappeDeprecationWarning"
+
+
+def _get_dynamic_base_classes(graduation: str) -> tuple[type, ...]:
+	# Scrub the graduation string to ensure it's a valid class name
+	import frappe
+
+	if version("frappe").upper().startswith(graduation.upper().removeprefix("V")):
+		return (FrappeDeprecationWarning,)
+	return (
+		FrappeDeprecationWarning,
+		PendingFrappeDeprecationWarning,
+	)
+
+
+def _get_dynamic_warning_class(class_name: str, base_classes: tuple[type, ...]):
+	current_module = sys.modules[__name__]
+	if hasattr(current_module, class_name):
+		return getattr(current_module, class_name)
+
+	dynamic_class = type(class_name, base_classes, {})
+	setattr(current_module, class_name, dynamic_class)
+	return dynamic_class
```

Comment:
I feel like this part would benefit from being less dynamic. If I understand correctly, this code is _monkey patching_ the deprecation_dumpster module based on the value of the `PYTHONWARNINGS` environment variable, which is fine.

If believe it would be more maintainable to make an explicit list of classes instead:

```python
class PendingFrappeDeprecationWarning(Warning):
  """Deprecated feature in develop"""

class FrappeDeprecationWarning(Warning):
  """Deprecated feature in current version"""

class FrappeDeprecationError(Exception):
  """
  We can go even further, and raise an error when the deprecated function is still used after its EOL.
  Raises an error when used, the body of the deprecated function can be removed, but its signature is kept.
  """

class V15FrappeDeprecationWarning(FrappeDeprecationError):
  pass
class V16FrappeDeprecationWarning(FrappeDeprecationWarning):
  pass
class V17FrappeDeprecationWarning(PendingFrappeDeprecationWarning):
  pass
```

Then, when bumping major versions, we can just change the base class of the corresponding versioned deprecation warning and be done.

Reply:
Nice, I like the concept of your suggestion!

Yeah, indeed the dynamic part is being meant to make a release close to effortless.

But then, again, a release manager needs to review this file in any case so the benefit of that intricate automation is quite marginal at the cost of readability.